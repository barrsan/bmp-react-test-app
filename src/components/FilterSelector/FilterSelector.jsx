import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';


class FilterSelector extends Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    options: PropTypes.array, // eslint-disable-line
    handler: PropTypes.func.isRequired,
  }

  static defaultProps = {
    options: [],
  };

  state = {
    value: '',
  }

  handleSelectChange = (e) => {
    e.persist();
    const { handler, type } = this.props;

    this.setState(prevProps => ({
      ...prevProps, value: e.target.value,
    }), () => {
      const { value } = this.state;
      handler({
        type,
        value,
      });
    });
  }

  render() {
    const { value } = this.state;
    const { options } = this.props;

    return (
      <Fragment>
        {
          options.length
            ? (
              <div className="filter-select-wrapper form-group">
                <select
                  className="form-control"
                  value={value}
                  onChange={this.handleSelectChange}
                >
                  <option value="">Все</option>
                  {
                    options.map(i => (
                      <option key={i} value={i}>{i}</option>
                    ))
                  }
                </select>
              </div>
            )
            : null
        }
      </Fragment>
    );
  }
}

export default FilterSelector;
