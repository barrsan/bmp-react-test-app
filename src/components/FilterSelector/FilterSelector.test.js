import React from 'react';
import { shallow } from 'enzyme';
import FilterSelector from './FilterSelector';


describe('FilterSelector component', () => {
  const mockSetExhibitsFilter = jest.fn();
  const initialState = {
    value: '',
  };

  describe('Options not empty', () => {
    const props = {
      type: 'place',
      options: ['v1', 'v2', 'v3'],
      handler: mockSetExhibitsFilter,
    };

    const wrapper = shallow(<FilterSelector {...props} />);

    test('renders correctly', () => expect(wrapper).toMatchSnapshot());

    test('renders without crashing', () => {
      expect(wrapper).toHaveLength(1);
    });

    test('initialize with initial state', () => {
      expect(wrapper.state()).toEqual(initialState);
    });

    describe('When typing into search input', () => {
      const value = 'v2';

      beforeEach(() => {
        wrapper.find('select').simulate('change', {
          persist: () => {},
          target: {
            value,
          },
        });
      });

      test('updates value field in state', () => {
        expect(wrapper.state().value).toEqual(value);
      });

      test('calls the props.handler', () => {
        expect(mockSetExhibitsFilter).toBeCalled();
      });
    });

    afterAll(() => {
      wrapper.setState(initialState);
    });
  });


  describe('Options is empty', () => {
    const props = {
      type: 'place',
      options: [],
      handler: mockSetExhibitsFilter,
    };

    const wrapper = shallow(<FilterSelector {...props} />);

    test('renders correctly', () => expect(wrapper).toMatchSnapshot());

    test('renders without crashing', () => {
      expect(wrapper).toHaveLength(1);
    });
  });
});
