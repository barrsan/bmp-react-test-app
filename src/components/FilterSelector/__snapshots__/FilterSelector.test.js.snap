// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSelector component Options is empty renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterSelector
    handler={[Function]}
    options={Array []}
    type="place"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": null,
    },
    "ref": null,
    "rendered": null,
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": null,
      },
      "ref": null,
      "rendered": null,
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`FilterSelector component Options not empty renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterSelector
    handler={[Function]}
    options={
        Array [
            "v1",
            "v2",
            "v3",
          ]
    }
    type="place"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        className="filter-select-wrapper form-group"
>
        <select
                className="form-control"
                onChange={[Function]}
                value=""
        >
                <option
                        value=""
                >
                        Все
                </option>
                <option
                        value="v1"
                >
                        v1
                </option>
                <option
                        value="v2"
                >
                        v2
                </option>
                <option
                        value="v3"
                >
                        v3
                </option>
        </select>
</div>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <select
          className="form-control"
          onChange={[Function]}
          value=""
>
          <option
                    value=""
          >
                    Все
          </option>
          <option
                    value="v1"
          >
                    v1
          </option>
          <option
                    value="v2"
          >
                    v2
          </option>
          <option
                    value="v3"
          >
                    v3
          </option>
</select>,
        "className": "filter-select-wrapper form-group",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option
              value=""
>
              Все
</option>,
            Array [
              <option
                value="v1"
>
                v1
</option>,
              <option
                value="v2"
>
                v2
</option>,
              <option
                value="v3"
>
                v3
</option>,
            ],
          ],
          "className": "form-control",
          "onChange": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Все",
              "value": "",
            },
            "ref": null,
            "rendered": "Все",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "v1",
            "nodeType": "host",
            "props": Object {
              "children": "v1",
              "value": "v1",
            },
            "ref": null,
            "rendered": "v1",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "v2",
            "nodeType": "host",
            "props": Object {
              "children": "v2",
              "value": "v2",
            },
            "ref": null,
            "rendered": "v2",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": "v3",
            "nodeType": "host",
            "props": Object {
              "children": "v3",
              "value": "v3",
            },
            "ref": null,
            "rendered": "v3",
            "type": "option",
          },
        ],
        "type": "select",
      },
      "type": "div",
    },
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          className="filter-select-wrapper form-group"
>
          <select
                    className="form-control"
                    onChange={[Function]}
                    value=""
          >
                    <option
                              value=""
                    >
                              Все
                    </option>
                    <option
                              value="v1"
                    >
                              v1
                    </option>
                    <option
                              value="v2"
                    >
                              v2
                    </option>
                    <option
                              value="v3"
                    >
                              v3
                    </option>
          </select>
</div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <select
            className="form-control"
            onChange={[Function]}
            value=""
>
            <option
                        value=""
            >
                        Все
            </option>
            <option
                        value="v1"
            >
                        v1
            </option>
            <option
                        value="v2"
            >
                        v2
            </option>
            <option
                        value="v3"
            >
                        v3
            </option>
</select>,
          "className": "filter-select-wrapper form-group",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option
                value=""
>
                Все
</option>,
              Array [
                <option
                  value="v1"
>
                  v1
</option>,
                <option
                  value="v2"
>
                  v2
</option>,
                <option
                  value="v3"
>
                  v3
</option>,
              ],
            ],
            "className": "form-control",
            "onChange": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Все",
                "value": "",
              },
              "ref": null,
              "rendered": "Все",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "v1",
              "nodeType": "host",
              "props": Object {
                "children": "v1",
                "value": "v1",
              },
              "ref": null,
              "rendered": "v1",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "v2",
              "nodeType": "host",
              "props": Object {
                "children": "v2",
                "value": "v2",
              },
              "ref": null,
              "rendered": "v2",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": "v3",
              "nodeType": "host",
              "props": Object {
                "children": "v3",
                "value": "v3",
              },
              "ref": null,
              "rendered": "v3",
              "type": "option",
            },
          ],
          "type": "select",
        },
        "type": "div",
      },
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
